### ISSUE TITLE
graph: consider changing semantics of Len() for graph.Iterator

### ISSUE THREAD

**Issue by kortschak** [gonum/gonum#627]
### Background
The graph node and edge iterators free the graph packages from the size limits of `[]graph.Node` and `[]graph.Edge` when querying graphs. The freedom to be able to perform graph operations on very large graphs that are backed by disk was one of the reasons for changing to `int64` ID values rather than `int` although at the time I was envisioning only working on subgraphs that had been pull from those larger backing stores.

With store-agnostic iterators we can conceivably work directly (though slowly) with nodes and edges that are not-memory resident with the exception that the iterator `Len` method must have a value to will fit inside an `int`. This won't work for 32 bit archs and will also not work for very large graphs. Given that the `Len` method is also used to give the user information about potential allocations either directly or via `graph.NodesOf` _etc._ it also doesn't work for very large graphs that may be materialised into a slice.

### Proposal
I suggest that we add a minor change to the semantics of the `Len` method's returned value.

> If the size/order of the graph is either unknown, too costly to calculate or likely to be too large to materialise (all loosely defined) then `Len` may return `-1`.
> The iterator's `Next` loop would however still be able to calculate the number of items.

An alternative to this is to add an `ok bool` to `Len` method, though it is unclear to me how that will help except in the case of 32 archs, and then we would also nee to change the return integer type to `int64` which I don't think is tenable.

### Potential impact of proposal

The only change to the graph packages would be an addition to the documentation for `graph.Iterator` on the `Len` method and a defensive check for negative length in `graph.NodesOf`, `graph.EdgesOf`, `graph.WeightedEdgesOf`, `graph.LinesOf` and `graph.WeightedLinesOf` after the relevant `*Slicer` type switch, returning a nil slice if the length is negative.

There are no types in the graph packages that would be affected by any of these changes.

I would do the work.

**Comment by tzneal** [comment]
+1 from a random user.  I needed a descendants iterator for a directed graph, and thought about cleaning it up and sending as a PR, but I didn't want to implement Len().

**Comment by kortschak** [comment]
Can you elaborate? "Didn't want to implement Len" is not a hugely compelling argument. In fact this is one thing I am concerned about - see second para just added to the impact section.

**Comment by tzneal** [comment]
In my mind, iterator is a loose concept to walk a series in a particular order.  The length doesn't come into it much, for the functions that require length they can just traverse the iterator and construct a list.

Suppose I implement a traversal that works using Next/Node without computing the list of nodes.  To then implement Len() correctly, I would have to rework the code so that either:
a) I walk the list when creating the iterator to determine the length
b) I walk the list when Len() is called to determine the length

**Comment by kortschak** [comment]
That's true, but they are also the container that is used to handle collections of nodes and edges from graphs into the analytical routines. These reoutines fall into two categories in the graph packages. Those that can just walk a graph and those that must have access to all the nodes. The way the second set do this is via `graph.NodesOf` (similar for edges). Under the proposal above, a graph that returns an interator with a negative length would internally have a `nil` `[]graph.Node` to work with and terminate early. This is OK, we just say that they won't work with indeterminate iterators.

However, as a part of the migration from using `[]graph.Node` as the container type to using `graph.Nodes` I basically wrapped most of the calls to `g.Nodes()` with `graph.NodesOf` so the code just continued to work. This can be worked away in many cases (the functions not in the second set above), but will currently also unnecessarily terminate early.

There is a third set (e.g. the traverse package types and `path.DijkstraFrom`) that just work properly without any change.

**Comment by kortschak** [comment]
This is not closed as there may be future changes (additions) that expand on this. For example, [the possibility of an optional `Size() int64` method for negative len iterators](https://github.com/gonum/gonum/pull/776#issuecomment-451642870).
